src/keystore.h

```diff
@@ -103,6 +115,43 @@ class CBasicKeyStore : public CKeyStore
     virtual bool RemoveWatchOnly(const CScript &dest);
     virtual bool HaveWatchOnly(const CScript &dest) const;
     virtual bool HaveWatchOnly() const;
+
+    bool AddSpendingKeyPaymentAddress(const libzcash::SpendingKey &key, const libzcash::PaymentAddress &address);
+    bool HaveSpendingKey(const libzcash::PaymentAddress &address) const
+    {
+        bool result;
+        {
+            LOCK(cs_KeyStore);
+            result = (mapSpendingKeys.count(address) > 0);
+        }
+        return result;
+    }
+    bool GetSpendingKey(const libzcash::PaymentAddress &address, libzcash::SpendingKey &keyOut) const
+    {
+        {
+            LOCK(cs_KeyStore);
+            SpendingKeyMap::const_iterator mi = mapSpendingKeys.find(address);
+            if (mi != mapSpendingKeys.end())
+            {
+                keyOut = mi->second;
+                return true;
+            }
+        }
+        return false;
+    }
+    void GetPaymentAddresses(std::set<libzcash::PaymentAddress> &setAddress) const
```

Comment:
This API appears to only ever be used so far in your other PR on note tracking, but I think we should be creating `ZCNoteDecryption` objects for each spending key and caching those for use in checking input notes, and keeping them consistent with the spending keys currently added to or imported to the wallet.